### PR TITLE
[Feature] Support for identity column in sql table

### PR DESCRIPTION
**Pull request by hshahconsulting**
## Changes

As Databricks supports identity columns, I have added support for identity columns in databricks_sql_table.
Closes [[FEATURE] Identity column specification on databricks_sql_table](https://github.com/databricks/terraform-provider-databricks/issues/3492)
## Tests
 added unit test that asserts a correct statement has been generated when an identity column type is specified

- [x] `make test` run locally
- [x] implement integration test
- [x] relevant change in `docs/` folder

### CODE REVIEW

File: catalog/resource_sql_table_test.go [hunk 1]
```diff
@@ -33,6 +33,36 @@ func TestResourceSqlTableCreateStatement_External(t *testing.T) {
...
+			},
+		},
+	}
+	stmt := ti.buildTableCreateStatement()
+	assert.Contains(t, stmt, "CREATE EXTERNAL TABLE `main`.`foo`.`bar`")
+	assert.Contains(t, stmt, "USING DELTA")
+	assert.Contains(t, stmt, "(`id` BIGINT GENERATED BY DEFAULT AS IDENTITY NOT NULL, `name`  NOT NULL COMMENT 'a comment')")
+	assert.Contains(t, stmt, "LOCATION 's3://ext-main/foo/bar1' WITH (CREDENTIAL `somecred`)")
+	assert.Contains(t, stmt, "COMMENT 'terraform managed'")
+}
```
Comment (hectorcast-db) [inline catalog/resource_sql_table_test.go:64]: Can you also add an integration test to this file?
https://github.com/databricks/terraform-provider-databricks/blob/main/internal/acceptance/sql_table_test.go#L13

**Review by hectorcast-db — CHANGES_REQUESTED** [review]
Hi @hshahconsulting 
Thanks for this contribution. I looks quite good. I left a couple of comment. 

@edwardfeng-db Can you also take a look at this PR?

File: catalog/resource_sql_table_test.go (hunk 1 above)
Comment (hshahconsulting) [reply catalog/resource_sql_table_test.go:64]: Done

**Review by edwardfeng-db — APPROVED** [review]

**Comment by hshahconsulting** [comment]
@hectorcast-db I have made the changes according to the feedback. Anything else that needs to be done? Also, @tanmay-db said that other changes may be needed. ny follow up on that?

**Comment by tanmay-db** [comment]
Hi @hshahconsulting, thanks for creating the PR. Can you please check if this works locally for you? You can create a local binary to test: https://github.com/databricks/terraform-provider-databricks/blob/main/CONTRIBUTING.md#developing-provider

So the steps would be to:
1. Create local binary (run `make` in root of repository over your PR changes)
2. Override the provider using ~/.terraformrc as mentioned in the link
3. Do terraform apply as you were doing before over your HCL. 

Note: Please use dev environment / test data for this (just in case there are some issues).

File: catalog/resource_sql_table.go [hunk 2]
```diff
@@ -24,6 +24,7 @@ var MaxSqlExecWaitTimeout = 50
 type SqlColumnInfo struct {
 	Name     string `json:"name"`
 	Type     string `json:"type_text,omitempty" tf:"alias:type,computed"`
+	Identity string `json:"identity,omitempty" tf:"default:false"`
```
Comment (hectorcast-db) [inline catalog/resource_sql_table.go:27]: We can use the IdentifyColumn here 
```
Identity IdentityColumn `json....`
```
and then there is no need to convert types below.
In that case, you will need to create a 3rd const below for the default case.

File: catalog/resource_sql_table_test.go (hunk 1 above)
Comment (hectorcast-db) [reply catalog/resource_sql_table_test.go:64]: Sorry for the missunderstanding: I was talking to add a test in the linked file so we can test this against the a real workspace.

**Review by hectorcast-db — COMMENTED** [review]
I see some issue in the list of commits. It seems that this PR is changing preexisting commits. Can you take a look at it?

File: catalog/resource_sql_table_test.go (hunk 1 above)
Comment (hshahconsulting) [reply catalog/resource_sql_table_test.go:64]: My bad, sorry. Will add.

**Comment by hshahconsulting** [comment]
> I see some issue in the list of commits. It seems that this PR is changing preexisting commits. Can you take a look at it?

I made a mistake with a previous rebase. The branch has now been cleaned up.

File: catalog/resource_sql_table.go (hunk 2 above)
Comment (hshahconsulting) [reply catalog/resource_sql_table.go:27]: Done

File: catalog/resource_sql_table_test.go (hunk 1 above)
Comment (hshahconsulting) [reply catalog/resource_sql_table_test.go:64]: I have added an integration test. However, I do not know how to run it.

**Comment by hshahconsulting** [comment]
> Hi @hshahconsulting, thanks for creating the PR. Can you please check if this works locally for you? You can create a local binary to test: https://github.com/databricks/terraform-provider-databricks/blob/main/CONTRIBUTING.md#developing-provider
> 
> So the steps would be to:
> 
> 1. Create local binary (run `make` in root of repository over your PR changes)
> 2. Override the provider using ~/.terraformrc as mentioned in the link
> 3. Do terraform apply as you were doing before over your HCL.
> 
> Note: Please use dev environment / test data for this (just in case there are some issues).

I am trying to run this locally but I cannot stop it from downloading the terraform provider from the internet. We have a version.tf with this code

`terraform {
  required_version = ">=1.7.1"
  required_providers {
    azurerm = {
      source  = "hashicorp/azurerm"
      version = "~> 3.89"
    }
    databricks = {
      source  = "databricks/databricks"
      version = "~>1.50"
    }
  }
  backend "azurerm" {
    resource_group_name  = "unity-tf-backend-rg"
    storage_account_name = "infradevseunitytfbackend"
    container_name       = "tf-backend"
    key                  = "terraform.tfstate"
  }
}
`

File: catalog/resource_sql_table_test.go (hunk 1 above)
Comment (hectorcast-db) [reply catalog/resource_sql_table_test.go:64]: I did run the tests and they succeed. Thanks!

Comment (hshahconsulting) [reply catalog/resource_sql_table_test.go:64]: Wohooo! Would you still like me to test locally or is this good enough? I have trouble of setting up what Tanmay suggested.

File: catalog/resource_sql_table.go [hunk 3]
```diff
@@ -188,7 +194,25 @@ func (ti *SqlTableInfo) getOrCreateCluster(clusterName string, clustersAPI clust
...
+	case IdentityColumnAlways:
+		identityType = "BIGINT GENERATED ALWAYS AS IDENTITY"
+	case IdentityColumnDefault:
+		identityType = "BIGINT GENERATED BY DEFAULT AS IDENTITY"
+	}
+	return identityType
+}
+
 func (ti *SqlTableInfo) serializeColumnInfo(col SqlColumnInfo) string {
+	var colType string
```
Comment (hectorcast-db) [inline catalog/resource_sql_table.go:209]: While testing this, I saw some cases where Identify was empty. This happens when SqlColumnInfo is build by this package (and not directly by TF). 
In such cases, the `colType` ends empty and will cause issues.
We can replace the `serializeIdentityColumnInfo` by a `getColumnType`

```
func (ci *SqlColumnInfo) getColumnType() string {
	var colType string
	switch ci.Identity {
	case IdentityColumnAlways:
		colType = "BIGINT GENERATED ALWAYS AS IDENTITY"
	case IdentityColumnDefault:
		colType = "BIGINT GENERATED BY DEFAULT AS IDENTITY"
        default:
                colType = col.Type
	}
	return colType

```

and then remove the if/else from here:
```
var colType = getColumnType()
```

Comment (hshahconsulting) [reply catalog/resource_sql_table.go]: Done

File: catalog/resource_sql_table_test.go (hunk 1 above)
Comment (hectorcast-db) [reply catalog/resource_sql_table_test.go:64]: No needed to.

**Review by hectorcast-db — COMMENTED** [review]
Content looks good. 
I just saw that there are issues with the git history again. Can you fix that? Then we can run the automatic workflows and merge.

File: catalog/resource_sql_table_test.go (hunk 1 above)
Comment (hectorcast-db) [reply catalog/resource_sql_table_test.go:64]: No need to since tests are passing. Thank!

**Comment by hshahconsulting** [comment]
> Content looks good. I just saw that there are issues with the git history again. Can you fix that? Then we can run the automatic workflows and merge.

I am not sure what the issue is this time. I created this branch from main and then pasted in the code. Which commit is bad?

**Comment by hshahconsulting** [comment]
> Content looks good. I just saw that there are issues with the git history again. Can you fix that? Then we can run the automatic workflows and merge.

Please check again @hectorcast-db

**Comment by hectorcast-db** [comment]
Hi @hshahconsulting 

I did run all integration tests and this changes causes some issues related to the usage of the `false` default:

* This test fails:
https://github.com/databricks/terraform-provider-databricks/blob/main/internal/acceptance/sql_table_test.go#L397
The given reason is that the `diff` after apply is not empty. 
```
 error=
  | After applying this test step, the plan was not empty.
  | stdout:
  | 
  | 
  | Terraform used the selected providers to generate the following execution
  | plan. Resource actions are indicated with the following symbols:
  |   ~ update in-place
  |  <= read (data resources)
  | 
  | Terraform will perform the following actions:
  | 
  |   # data.databricks_table.this will be read during apply
  |   # (depends on a resource or a module with changes pending)
  |  <= data "databricks_table" "this" {
  |       + id   = (known after apply)
  |       + name = "sandbox########.bar"
  |     }
  | 
  |   # databricks_sql_table.mytable will be updated in-place
  |   ~ resource "databricks_sql_table" "mytable" {
  |         id                 = "sandbox########.bar"
  |         name               = "bar"
  |         # (7 unchanged attributes hidden)
  | 
  |       ~ column {
  |           + identity = "false"
  |             name     = "id"
  |             # (3 unchanged attributes hidden)
  |         }
  |     }
  | 
  | Plan: 0 to add, 1 to change, 0 to destroy.
```
* This change causes a diff on all existing resources, which forces customers to update all their tables. Not a great experience, and we should avoid it if possible.

Making the default an empty string should solve both issues.

**Comment by hshahconsulting** [comment]
> Hi @hshahconsulting
> 
> I did run all integration tests and this changes causes some issues related to the usage of the `false` default:
> 
> * This test fails:
>   https://github.com/databricks/terraform-provider-databricks/blob/main/internal/acceptance/sql_table_test.go#L397
>   The given reason is that the `diff` after apply is not empty.
> 
> ```
>  error=
>   | After applying this test step, the plan was not empty.
>   | stdout:
>   | 
>   | 
>   | Terraform used the selected providers to generate the following execution
>   | plan. Resource actions are indicated with the following symbols:
>   |   ~ update in-place
>   |  <= read (data resources)
>   | 
>   | Terraform will perform the following actions:
>   | 
>   |   # data.databricks_table.this will be read during apply
>   |   # (depends on a resource or a module with changes pending)
>   |  <= data "databricks_table" "this" {
>   |       + id   = (known after apply)
>   |       + name = "sandbox########.bar"
>   |     }
>   | 
>   |   # databricks_sql_table.mytable will be updated in-place
>   |   ~ resource "databricks_sql_table" "mytable" {
>   |         id                 = "sandbox########.bar"
>   |         name               = "bar"
>   |         # (7 unchanged attributes hidden)
>   | 
>   |       ~ column {
>   |           + identity = "false"
>   |             name     = "id"
>   |             # (3 unchanged attributes hidden)
>   |         }
>   |     }
>   | 
>   | Plan: 0 to add, 1 to change, 0 to destroy.
> ```
> 
> * This change causes a diff on all existing resources, which forces customers to update all their tables. Not a great experience, and we should avoid it if possible.
> 
> Making the default an empty string should solve both issues.

Is this correct?

	Identity IdentityColumn `json:"identity,omitempty"`
	
I have never coded in Go Lang before so please excuse my beginner level questions hehe :D

**Comment by hshahconsulting** [comment]
Please check. I have set the default to empty string.

**Comment by hshahconsulting** [comment]
@hectorcast-db Sorry for bumping but anything else needs fixing?

**Comment by hectorcast-db** [comment]
Tests just finished, and there is still one issue. However, I don't know the API well enough to recommend a solution. I have contacted the service team and I am waiting for a response. I will describe the issue in case you know how to fix it:

The new test you added is failing. Terraform always reads the server data after an apply, and then validates that there is no diff between the server status and the local status. 

However, the server does not return the Identity, so there is always a diff.

We need to be able to reconstruct the "Identity" value inside the READ method based on the server response.

**Comment by hshahconsulting** [comment]
@hectorcast-db Hello Hector. Do you have any news regarding the Service Team planning to take a look at this?

**Comment by hshahconsulting** [comment]
@hectorcast-db thanks. Could you paste the logs from the latest run of the integration tests where this fails, after I set the default value for the identity to an empty string? Perhaps I could figure something out by looking at it.

**Comment by hshahconsulting** [comment]
Duplicate of #4035

**Pull request closed without merging**